### DISCUSSION TITLE
Using Key Vault for X509CertificateAuthenticationProvider

### DISCUSSION THREAD

**Discussion by cbarnes01** [pnp/pnpcore#514] [Q&A]
Can you provide guidance on using the Azure Key Vault to store certificates for use with X509CertificateAuthenticationProvider.

**Comment by jansenbe** [comment]
You can combine KeyVault with PnP Core SDK: when using certificates in PnP Core SDK either you provide a pointer to the certificate by specifying the store name (e.g. My), the location (e.g. LocalMachine) and the thumbprint of the cert or alternatively you provide the actual certificate as an `X509Certificate2`.  They KeyVault .NET API does have a method to get an `X509Certificate2`: https://docs.microsoft.com/en-us/dotnet/api/azure.security.keyvault.certificates.certificateclient.downloadcertificateasync?view=azure-dotnet#Azure_Security_KeyVault_Certificates_CertificateClient_DownloadCertificateAsync_System_String_System_String_System_Threading_CancellationToken_ and https://docs.microsoft.com/en-us/azure/key-vault/certificates/quick-create-net

**Reply by juliemturner** [reply]
So what I think I'm reading is that before I would just (for local development purposes) export the thumbprint of the Azure Key Vault certificate and use that via my local.settings.json and then when deployed, for the Azure Function, it could be integrated into the Application Settings... I'm gathering (based on trying a few different things and having no luck) that that's not going to work...

**Reply by cbarnes01** [reply]
Julie,
This may not be relevant for your situation, but I found this solution works without having to use the Key Vault.
https://github.com/pnp/pnpcore/tree/dev/samples/Demo.AzureFunction

**Reply by juliemturner** [reply]
That's the sample I'm using... but I want to use Key Vault so sadly I'm still trying to figure it out. Thanks.

**Reply by cbarnes01** [reply]
Julie,
I was struggling with this too, this is the configuration I was able to get working.

Startup.cs
```
public class Startup : FunctionsStartup
   {
        public override void Configure(IFunctionsHostBuilder builder)
        {
            builder.Services.AddPnPCore((options) =>
             {
                 options.DisableTelemetry = true;

                 var authProvider = new X509CertificateAuthenticationProvider(
                     Environment.GetEnvironmentVariable("ClientId"),
                     Environment.GetEnvironmentVariable("TenantId"),
                     new PnPCoreAuthenticationX509CertificateOptions()
                     {
                         Certificate = GetCertificate(new Uri(Environment.GetEnvironmentVariable("KeyVaultUrl")),
                             Environment.GetEnvironmentVariable("CertName"))
                     }
                 );
                 options.DefaultAuthenticationProvider = authProvider;

                 options.Sites.Add("SharePointSite", new PnPCoreSiteOptions
                 {
                     SiteUrl = Environment.GetEnvironmentVariable("SharePointUrl"),
                     AuthenticationProvider = authProvider
                 });
             });
        }

        private static X509Certificate2 GetCertificate(Uri keyVaultUrl, string certName)
        {
            var client = new CertificateClient(keyVaultUrl, new DefaultAzureCredential());
            var cert = client.DownloadCertificate(certName);

            return cert;
        }
    }
```

**Reply by juliemturner** [reply]
Yes, I see, and I get it but that code would not be needed in the Azure Function and I was trying to avoid having to have code done one way for testing and a different way for production, although I get that you could use this code for production it's not the way I think is best done for Azure Functions, it's easier and more obvious to be configured as an app setting.  

I have solved it at some level... but I admit I have no idea why it's working.  Basically, after I uploaded my cert to the app registration providing the "thumbprint" that's in the app registration to the auth provider seems to work.. I'm getting data... I need to test a few more things but at this point at least I can move forward.

**Reply by PaoloPia** [reply]
If needed, we can eventually slightly change the provider to support Azure Key Vault, too ... let's talk about it ...

**Reply by cbarnes01** [reply]
Julie,
Does Environment.GetEnvironmentVariable not pull from the Application settings  in the function app?  Sorry I've only testing this locally at this point.  Maybe that's not what you are referring to?

**Reply by juliemturner** [reply]
It does, but the value you're putting there would not be needed... you can get the certificate itself right in the variable. so there would be no need for the URL or CertName

**Reply by cbarnes01** [reply]
Julie,
Sorry trying to understand.  Are you saying to store the cert in an application setting variable?

**Reply by juliemturner** [reply]
No... i'm saying you reference the cert in the app setting...  I also store things like ClientId, TenantId, etc...

`@Microsoft.KeyVault(SecretUri=https://{MYCERTPATH})`

**Reply by cbarnes01** [reply]
Julie,
I see, your configuration must use Microsoft.Extensions.Configuration.AzureKeyVault or the package that replaced it.  Would you mind posting your startup file?  Thanks!

**Reply by juliemturner** [reply]
Nope, it doesn't... it just gets the certificate from the environmental variable just like any other app setting...

`private static string Cert = Environment.GetEnvironmentVariable("CertificateFromKeyVault");`

**Reply by cbarnes01** [reply]
Julie,
Would love to see your completed configuration, I've just started using the Key Vault and I'm sure it would be informative.
